File: java/code/49.java
import java.util.*;

class Solution {
    public String sort(String s){
        char[] chars = s.toCharArray();
        Arrays.sort(chars);
        return new String(chars);
    }
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> ans = new ArrayList();
        HashMap<String, List<String>> mp = new HashMap<>();
        for (String s: strs){
            List<String> list = mp.get(sort(s));
            if(list == null){
                list = new ArrayList<>();
                list.add(s);
                mp.put(sort(s), list);
            }
            else{
                list.add(s);
            }
        }
        for (Map.Entry<String, List<String>> entry: mp.entrySet()){
            List<String> value = entry.getValue();
            ans.add(value);
        }
        return ans;
    }
}

File: C++/code/49.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> ans;
        unordered_map<string, vector<string>> mp;
        for (string s: strs) {
            vector<int> cnt(26, 0);
            for (char ch: s) cnt[ch - 'a']++;
            string key;
            for (int i = 0; i < 26; i++) {
                for (int j = 0; j < cnt[i]; j++) {
                    key += 'a' + i;
                }
            }
            mp[key].push_back(s);
        }
        for (auto& pair: mp) {
            ans.push_back(pair.second);
        }
        return ans;
    }
};